File: mathlib/expressions/Exponent.ts
import { inParen, inRow } from "../util/MathMLHelpers";
import { Expression } from "./Expression";
import { Integer, IntegerType } from "./Integer";
import { ProductType } from "./Product";
import { SumType } from "./Sum";


export class Exponent extends Expression {
    public static of(base: Expression, power: Expression): Exponent {
        const hash = base.hash + power.hash
        if (!Exponent.instances.has(hash)) {
            Exponent.instances.set(hash, new Exponent(base, power))
        }
        return Exponent.instances.get(hash)!
    }
    private static instances = new Map<string, Exponent>()

    public class: string = ExponentType;
    public toMathXML(): string {
        function wrapIfNeeded(exp: Expression): string {
            if (exp.class == SumType || exp.class == ProductType)
                return inRow(inParen(exp.toMathXML()))
            return exp.toMathXML()
        }
        return "<msup>" + wrapIfNeeded(this.base) + this.power.toMathXML() + "</msup>"
    }
    public toString(): string {
        return "(" + this.base + ")^(" + this.power + ")"
    }

    public get hash(): string {
        return "Exponent"+this.base.hash+this.power.hash
    }

    private constructor(base: Expression, power: Expression) {
        super()
        this.base = base
        this.power = power
        Object.freeze(this.base)
        Object.freeze(this.power)
        // The integers are closed over exponentiation
        this.isReducible = (base.isReducible || base.class == IntegerType) && (power.isReducible || power.class == IntegerType) && Math.pow(base.reduced.value, power.reduced.value) % 1 == 0
        this.isHealthy = !this.isReducible
        this.isConstant = base.isConstant && power.isConstant
        this.childCount = 2 + base.childCount + power.childCount
    }
    
    public readonly base: Expression;
    public readonly power: Expression;
    public readonly isReducible: boolean;
    public readonly isHealthy: boolean;

    public get reduced(): Integer {
        return Integer.of(Math.pow(this.base.reduced.value, this.power.reduced.value))
    }
    public readonly isConstant: boolean;
    public readonly childCount: number;
} 

export const ExponentType = "Exponent"

File: cas/src/expressions/exponent.rs
use serde_json::json;

use crate::mathxml::{in_row, in_paren};

use super::{ExpressionPtr, IExpression, Expression, EXPRESSION_INSTANCES};
use core::fmt;
use std::sync::Arc;

#[derive(PartialEq, Eq, Hash)]
pub struct Exponent {
    base: ExpressionPtr,
    power: ExpressionPtr,
}

impl Exponent {
    pub fn of(base: ExpressionPtr, power: ExpressionPtr) -> ExpressionPtr {
        let id = get_id(&base, &power);

        let mut instances = EXPRESSION_INSTANCES.lock().unwrap();

        if let Some(result) = instances.get(&id) {
            return result.clone();
        }

        let result = Expression::Exponent(Arc::new(Exponent{base, power}));
        instances.insert(id, result.clone());
        result
    }

    pub fn base(&self) -> ExpressionPtr {
        self.base.clone()
    }

    pub fn power(&self) -> ExpressionPtr {
        self.power.clone()
    }
}

impl IExpression for Exponent {
    fn to_unambigious_string(&self) -> String {
        format!("({})^({})", 
            self.base.as_stringable().to_unambigious_string(), 
            self.power.as_stringable().to_unambigious_string())
    }

    fn to_math_xml(&self) -> String {
        fn wrap_if_needed(exp: &ExpressionPtr) -> String {
            let stringable = exp.as_stringable();
            match exp {
                Expression::Sum(s) =>
                    in_row(&in_paren(&s.to_math_xml())),
                Expression::Product(p) => 
                    in_row(&in_paren(&p.to_math_xml())),
                _ => stringable.to_math_xml(),
            }
        }

        format!(
            "<msup>{}{}</msup>",
            wrap_if_needed(&self.base),
            &in_row(&self.power.as_stringable().to_math_xml())
        )
    }

    fn id(&self) -> String {
        get_id(&self.base, &self.power)
    }

    fn to_json(&self) -> serde_json::Value {
        json!([
            "Pow",
            self.base.to_json(),
            self.power.to_json()
        ])
    }
}

fn get_id(base: &ExpressionPtr, power: &ExpressionPtr) -> String {
        format!(
            "exponent{}{}",
            &base.as_stringable().id(),
            &power.as_stringable().id()
        )
}

impl fmt::Debug for Exponent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Exponent({}, {})", self.base, self.power)
    }
}

#[cfg(test)]
mod tests {
    use crate::expressions::Integer;

    use super::Exponent;

    #[test]
    fn flywheel_impl() {

        let first = Exponent::of(Integer::of(1), Integer::of(1));
        let second = Exponent::of(Integer::of(1), Integer::of(1));
        let third = Exponent::of(Integer::of(1), Integer::of(0));
        let forth = Exponent::of(Integer::of(0), Integer::of(1));

        assert_eq!(first, second);
        assert_eq!(first.as_stringable().id(), second.as_stringable().id());
        assert_ne!(first, third);
        assert_ne!(first, forth);
        assert_ne!(third, forth);
    }

}
